File: src/Tiedragon.XmppMessenger.Core/Xmpp/XmppServiceContactAddresses.cs
using System.Xml.Linq;

namespace Tiedragon.XmppMessenger.Core.Xmpp;

public static class XmppServiceContactAddresses
{
    public const string FormType = "http://jabber.org/network/serverinfo";

    public const string AbuseField = "abuse-addresses";

    public const string AdminField = "admin-addresses";

    public const string FeedbackField = "feedback-addresses";

    public const string SalesField = "sales-addresses";

    public const string SecurityField = "security-addresses";

    public const string StatusField = "status-addresses";

    public const string SupportField = "support-addresses";

    private static readonly XmppServiceContactAddressKind[] OrderedKinds =
    [
        XmppServiceContactAddressKind.Abuse,
        XmppServiceContactAddressKind.Admin,
        XmppServiceContactAddressKind.Feedback,
        XmppServiceContactAddressKind.Sales,
        XmppServiceContactAddressKind.Security,
        XmppServiceContactAddressKind.Status,
        XmppServiceContactAddressKind.Support
    ];

    public static bool TryGetContactAddresses(
        XmppServiceDiscoveryInfo info,
        out IReadOnlyList<XmppServiceContactAddress> addresses)
    {
        ArgumentNullException.ThrowIfNull(info);

        var parsed = new List<XmppServiceContactAddress>();
        foreach (var form in info.DataForms)
        {
            if (!string.Equals(form.FormType, FormType, StringComparison.Ordinal))
            {
                continue;
            }

            foreach (var kind in OrderedKinds)
            {
                if (!form.Fields.TryGetValue(FieldName(kind), out var values))
                {
                    continue;
                }

                foreach (var value in values)
                {
                    if (Uri.TryCreate(value, UriKind.Absolute, out var uri))
                    {
                        parsed.Add(new XmppServiceContactAddress(kind, uri));
                    }
                }
            }
        }

        addresses = parsed
            .DistinctBy(address => (address.Kind, address.Uri.OriginalString), EqualityComparer<(XmppServiceContactAddressKind, string)>.Default)
            .ToArray();
        return addresses.Count > 0;
    }

    public static IReadOnlyList<Uri> GetAddresses(
        XmppServiceDiscoveryInfo info,
        XmppServiceContactAddressKind kind)
    {
        return TryGetContactAddresses(info, out var addresses)
            ? addresses.Where(address => address.Kind == kind).Select(address => address.Uri).ToArray()
            : [];
    }

    public static XElement CreateDataForm(IEnumerable<XmppServiceContactAddress> addresses)
    {
        ArgumentNullException.ThrowIfNull(addresses);

        var fields = addresses
            .GroupBy(address => address.Kind)
            .OrderBy(group => Array.IndexOf(OrderedKinds, group.Key))
            .Select(group => new XElement(
                XName.Get("field", XmppServiceDiscovery.DataFormNamespace),
                new XAttribute("var", FieldName(group.Key)),
                group.Select(address => new XElement(
                    XName.Get("value", XmppServiceDiscovery.DataFormNamespace),
                    address.Uri.OriginalString))));

        return new XElement(
            XName.Get("x", XmppServiceDiscovery.DataFormNamespace),
            new XAttribute("type", "result"),
            new XElement(
                XName.Get("field", XmppServiceDiscovery.DataFormNamespace),
                new XAttribute("var", "FORM_TYPE"),
                new XAttribute("type", "hidden"),
                new XElement(XName.Get("value", XmppServiceDiscovery.DataFormNamespace), FormType)),
            fields);
    }

    public static string FieldName(XmppServiceContactAddressKind kind)
    {
        return kind switch
        {
            XmppServiceContactAddressKind.Abuse => AbuseField,
            XmppServiceContactAddressKind.Admin => AdminField,
            XmppServiceContactAddressKind.Feedback => FeedbackField,
            XmppServiceContactAddressKind.Sales => SalesField,
            XmppServiceContactAddressKind.Security => SecurityField,
            XmppServiceContactAddressKind.Status => StatusField,
            XmppServiceContactAddressKind.Support => SupportField,
            _ => throw new ArgumentOutOfRangeException(nameof(kind))
        };
    }
}

public enum XmppServiceContactAddressKind
{
    Abuse,
    Admin,
    Feedback,
    Sales,
    Security,
    Status,
    Support
}

public sealed record XmppServiceContactAddress(
    XmppServiceContactAddressKind Kind,
    Uri Uri);

File: php/lib/Xmpp/XmppServiceContactAddresses.php
<?php
declare(strict_types=1);

namespace Tiedragon\Xmpp;

final class XmppServiceContactAddresses
{
    /** @var array<string,string> */
    private const FIELD_BY_KIND = [
        'abuse' => 'abuse-addresses',
        'admin' => 'admin-addresses',
        'feedback' => 'feedback-addresses',
        'sales' => 'sales-addresses',
        'security' => 'security-addresses',
        'status' => 'status-addresses',
        'support' => 'support-addresses',
    ];

    /**
     * @param array<string,list<string>|string> $addresses
     */
    public static function dataFormElement(array $addresses): string
    {
        $fields = ['FORM_TYPE' => ['type' => 'hidden', 'value' => XmppXml::SERVERINFO_NS]];
        foreach (self::FIELD_BY_KIND as $kind => $fieldName) {
            if (!array_key_exists($kind, $addresses)) {
                continue;
            }

            $values = is_array($addresses[$kind]) ? $addresses[$kind] : [$addresses[$kind]];
            $values = array_values(array_filter(array_map('strval', $values), self::isAbsoluteUri(...)));
            if ($values !== []) {
                $fields[$fieldName] = ['values' => $values, 'type' => 'list-multi'];
            }
        }

        return XmppDataForm::formElement($fields, 'result');
    }

    /**
     * @return list<array{kind:string,uri:string}>
     */
    public static function parseFromDiscoInfo(string $xml): array
    {
        $contacts = [];
        foreach (XmppDataForm::parseForms($xml) as $form) {
            $formType = $form['fields']['FORM_TYPE']['values'][0] ?? null;
            if ($formType !== XmppXml::SERVERINFO_NS) {
                continue;
            }

            foreach (self::FIELD_BY_KIND as $kind => $fieldName) {
                foreach ($form['fields'][$fieldName]['values'] ?? [] as $uri) {
                    if (self::isAbsoluteUri($uri)) {
                        $contacts[$kind . "\n" . $uri] = ['kind' => $kind, 'uri' => $uri];
                    }
                }
            }
        }

        return array_values($contacts);
    }

    private static function isAbsoluteUri(string $value): bool
    {
        return preg_match('/^[A-Za-z][A-Za-z0-9+.-]*:/', $value) === 1;
    }
}
